### PR TITLE
Update to the HarmonicOscillatorWaveFunction with a wave function recurrence that uses Cython.

### PR DESCRIPTION
**Pull request by fobos123deimos**
**Checklist**
Thank you for contributing to QuTiP! Please make sure you have finished the following tasks before opening the PR.

- [x] Please read [Contributing to QuTiP Development](http://qutip.org/docs/latest/development/contributing.html)
- [x] Contributions to qutip should follow the [pep8 style](https://www.python.org/dev/peps/pep-0008/).
You can use [pycodestyle](http://pycodestyle.pycqa.org/en/latest/index.html) to check your code automatically
- [x] Please add tests to cover your changes if applicable.
- [ ] If the behavior of the code has changed or new feature has been added, please also update the documentation in the `doc` folder, and the [notebook](https://github.com/qutip/qutip-tutorials). Feel free to ask if you are not sure.
- [x] Include the changelog in a file named: `doc/changes/<PR number>.<type>` 'type' can be one of the following: feature, bugfix, doc, removal, misc, or deprecation (see [here](http://qutip.org/docs/latest/development/contributing.html#changelog-generation) for more information).

Delete this checklist after you have completed all the tasks. If you have not finished them all, you can also open a [Draft Pull Request](https://github.blog/2019-02-14-introducing-draft-pull-requests/) to let the others know this on-going work and keep this checklist in the PR description.

**Description**

I am adding an efficient module for calculating the wave function that uses a recurrence for it in Cython, and I am applying one of its functions in the update method of the HarmonicOscillatorWaveFunction class in the distribution.py module. This module is an adaptation of one of the modules from the package I developed for efficient wave function calculation: Fast Wave.

**Related issues or PRs**
Please mention the related issues or PRs here. If the PR fixes an issue, use the keyword fix/fixes/fixed followed by the issue id, e.g. fix #1184

### CODE REVIEW

**Comment by fobos123deimos** [comment]
I will resolve the issues!

Matheus.

**Comment by fobos123deimos** [comment]
@hodgestar

File: qutip/distributions.py [hunk 1]
```diff
@@ -357,6 +358,20 @@ def update_rho(self, rho):
...
+    Cython version of QuTip, from the wavefunction_cython
+    module of the Fast_Wave <https://github.com/fobos123deimos/fast-wave>
+    package, which is released under
+    the BSD license, with the following copyright notice:
+    CORDEIRO, MATHEUS; BEZERRA, I. P. ; VASCONCELOS, H. H. M. . Efficient
+    Computation of the Wave Function ψn(x) using Hermite Coefficient Matrix in
+    Python. In: 7º Workshop Escola de Computação e Informação Quântica
+    (7ª WECIQ),2024, Rio de Janeiro.ANAIS DO 7º WORKSHOP-ESCOLA DE COMPUTAÇÃO
+    E INFORMAÇÃO QUÂNTICA. Rio de Janeiro: CEFET/RJ, 2024. p. 56-60.
+    """
```
Comment (Ericgig) [inline qutip/distributions.py:373]: This docstring is not added to the `Distribution` one, but replace it. 
So it should have a description of what it is, it's parameters etc.

It's fine to add reference to your paper, but that not as the only description of the class.

File: qutip/wavefunction/wavefunction_cython.pyx [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
```
Comment (Ericgig) [inline qutip/wavefunction/wavefunction_cython.pyx:30]: Would you be comfortable to be simply included as "QuTiP developers and contributors" of the main BSD 3-Clause License of Qutip?

Having you as the author of the file give the feeling that no one else should touch it and we don't want that.
If merged, we will eventually need to modify this file, maybe merge it or split it etc.

File: qutip/wavefunction/wavefunction_cython.pyx [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+            for k in range(x_size):
+                result[i + 1, k] = 2 * x[k] * (result[i, k] / temp1) - temp2 * result[i - 1, k]
+
+    return result[-1, :]
+
+@cython.nogil
+@cython.cfunc
+@cython.locals(x_size=np.npy_intp, j=int, i=int, k=int, temp1=complex, temp2=complex)
+@cython.boundscheck(False)
+cpdef np.ndarray[np.complex128_t, ndim=1] psi_n_single_fock_multiple_position_complex(int n, np.ndarray[np.complex128_t, ndim=1] x):
```
Comment (Ericgig) [inline qutip/wavefunction/wavefunction_cython.pyx:206]: This is the only function used...
Can you remove the others?

File: qutip/wavefunction/wavefunction_cython.pyx [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+cimport cython
+from cython cimport double, complex
+cimport numpy as np
+import numpy as np
+from libc.math cimport exp, sqrt, pi
+from cmath import exp as cexp, sqrt as csqrt, pi as cpi
```
Comment (Ericgig) [inline qutip/wavefunction/wavefunction_cython.pyx:37]: To import the c++ versions:

```suggestion
cdef extern from "<complex>" namespace "std" nogil:
    double complex cexp "exp"(double complex x)
    double complex csqrt "sqrt"(double complex x)
```

File: qutip/wavefunction/wavefunction_cython.pyx [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+        else:
+            for k in range(x_size):
+                result[i + 1, k] = 2 * x[k] * (result[i, k] / temp1) - temp2 * result[i - 1, k]
+
+    return result[-1, :]
+
+@cython.nogil
+@cython.cfunc
+@cython.locals(x_size=np.npy_intp, j=int, i=int, k=int, temp1=complex, temp2=complex)
+@cython.boundscheck(False)
```
Comment (Ericgig) [inline qutip/wavefunction/wavefunction_cython.pyx:205]: You could add `@cython.cdivision(True)` to run in pure cython.

File: qutip/wavefunction/wavefunction_cython.pyx [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+    ----------
+    - Pérez-Jordá, J. M. (2017). On the recursive solution of the quantum harmonic oscillator. *European Journal of Physics*, 39(1), 
+      015402. doi:10.1088/1361-6404/aa9584
+    """
+    
+    x_size = x.shape[0]
+    cdef np.ndarray[np.complex128_t, ndim=2] result = np.zeros((n + 1, x_size), dtype=np.complex128)
+
+    for j in range(x_size):
+        result[0, j] = (cpi ** (-0.25)) * cexp(-(x[j] ** 2) / 2)
```
Comment (Ericgig) [inline qutip/wavefunction/wavefunction_cython.pyx:243]: If you want to optimize as much as possible, you could pre-compute the power `temp1 = cpi ** -0.25 / 2`.
But I wouldn't be surprised the compiler do it with the optimization we use (-O3).

**Review by Ericgig — COMMENTED** [review]
Thank you for the contribution.

Could you add a towncrier entry? 
Some tests would also be nice. I can't easily tell if you are computing the same thing looking at the code (not that I can confirm there is no bug in the original version.)
Also it would be better to have the cython code in `qutip/_distribution.pyx`. A new folder for one file feels overkill.

File: qutip/distributions.py [hunk 7]
```diff
@@ -376,12 +391,11 @@ def update(self, psi):
...
 
         for n in range(N):
-            k = pow(self.omega / pi, 0.25) / \
-                sqrt(2 ** n * factorial(n)) * \
-                exp(-self.xvecs[0] ** 2 / 2.0) * \
-                np.polyval(hermite(n), self.xvecs[0])
-
-            self.data += k * psi.data_as()[n, 0]
+            self.data += (
+                wc.psi_n_single_fock_multiple_position_complex(
```
[the remark this answers is not in the document]
Comment (fobos123deimos) [reply qutip/distributions.py]: I solved it.

File: qutip/distributions.py (hunk 1 above)
Comment (fobos123deimos) [reply qutip/distributions.py]: I solved it

File: qutip/wavefunction/wavefunction_cython.pyx (hunk 2 above)
Comment (fobos123deimos) [reply qutip/wavefunction/wavefunction_cython.pyx]: I solved it

File: qutip/wavefunction/wavefunction_cython.pyx (hunk 3 above)
Comment (fobos123deimos) [reply qutip/wavefunction/wavefunction_cython.pyx]: I solved it

File: qutip/wavefunction/wavefunction_cython.pyx (hunk 4 above)
Comment (fobos123deimos) [reply qutip/wavefunction/wavefunction_cython.pyx]: I had trouble solving this, so I left it as is.

File: qutip/wavefunction/wavefunction_cython.pyx (hunk 5 above)
Comment (fobos123deimos) [reply qutip/wavefunction/wavefunction_cython.pyx]: I solved it

File: qutip/wavefunction/wavefunction_cython.pyx (hunk 6 above)
Comment (fobos123deimos) [reply qutip/wavefunction/wavefunction_cython.pyx]: I solved it

File: qutip/distributions.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (fobos123deimos) [reply qutip/distributions.py]: I created the _distribution.pyx method, I will provide tests and towncrier entries.

**Review by Ericgig — COMMENTED** [review]
Thank you for the fixes and for writing a docstrings

The wave function generated does not match what I expected.
Here is what I got for the probability of the 5 first states:
download
ht?

Could add tests to check it's right.

File: qutip/distributions.py [hunk 8]
```diff
@@ -20,6 +20,7 @@
 
 from . import isket, ket2dm, state_number_index
 from .wigner import wigner, qfunc
+from _distributions import psi_n_single_fock_multiple_position_complex
```
[the remark this answers is not in the document]
Comment (fobos123deimos) [reply qutip/distributions.py:23]: Thank you! Could you provide your test code so I can use it as a reference to find the problem with this issue?

[the remark this answers is not in the document]
Comment (fobos123deimos) [reply qutip/distributions.py:23]: I managed to identify the error.

[the remark this answers is not in the document]
Comment (fobos123deimos) [reply qutip/distributions.py:23]: The problem was the decorator @cython.cdivision(True). I removed it.

[the remark this answers is not in the document]
Comment (fobos123deimos) [reply qutip/distributions.py:23]: I will do the tests, but is it ok to remove cdivision(True)?

[the remark this answers is not in the document]
Comment (Ericgig) [reply qutip/distributions.py:23]: Yes, it's ok.

[the remark this answers is not in the document]
Comment (fobos123deimos) [reply qutip/distributions.py]: I added the changelog, created the test script, named it `test_distributions.py`, and placed it in the `tests` folder. I also adjusted some parts of the `distributions.py` script that were causing bugs. I need to know where I should update the `doc` folder to reflect the changes I made. Additionally, which notebooks in the `qutip-tutorial` could I use to demonstrate my modifications?

**Review by Ericgig — APPROVED** [review]
Thank you. I would be happy to merge as is.
If you are fine with it, change the PR from draft to ready to review.

If you want to make a documentation page, leave it as draft and I will be waiting for it.

**Comment by fobos123deimos** [comment]
At the moment I'm not going to make a documentation page, but I intend to, including a very detailed one.

**Comment by fobos123deimos** [comment]
I changed the PR to ready for review.

**Pull request merged**